src/entity/Entity.php

```diff
@@ -1128,11 +1130,41 @@ public function isUnderwater() : bool{
 	}
 
 	public function isInsideOfSolid() : bool{
-		$block = $this->getWorld()->getBlockAt((int) floor($this->location->x), (int) floor($y = ($this->location->y + $this->getEyeHeight())), (int) floor($this->location->z));
+		$block = $this->getWorld()->getBlockAt((int) floor($this->location->x), (int) floor($this->location->y + $this->getEyeHeight()), (int) floor($this->location->z));
 
 		return $block->isSolid() && !$block->isTransparent() && $block->collidesWithBB($this->getBoundingBox());
 	}
 
+	/**
+	 * Checks if this entity is inside of a specific liquid
+	 * @param bool $checkEyePos If true, only the block at the eye position is checked for, otherwise all colliding liquids
+	 * @param bool $submerged   If true, all blocks in the bounding box must be the liquid
+	 */
+	public function isInsideOfLiquid(Liquid $liquid, bool $checkEyePos, bool $submerged = false) : bool{
+		$y = $this->location->y + $this->getEyeHeight();
+		$blockY = floor($y);
+		if($checkEyePos){
+			return $this->getWorld()->getBlockAt((int) floor($this->location->x), (int) $blockY, (int) floor($this->location->z)) instanceof $liquid;
+		}
+		$collisionBlocks = $this->getBlocksAroundWithEntityInsideActions();
+		$blocks = array_filter($collisionBlocks, static function($block) use ($liquid, $y, $blockY, $submerged) : bool{
+			if($submerged && $block instanceof $liquid){
+				$f = ($blockY + 1) - ($block->getFluidHeightPercent() - 0.1111111);
```

Comment:
What's this value 0.1111 ?

Reply:
https://github.com/pmmp/PocketMine-MP/blob/77db7a8941d3d91c7e452cb3ca469f59fbd8f4df/src/entity/Entity.php#L1123